tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/api/OpenPaasRestClient.java

```diff
@@ -0,0 +1,59 @@
+package com.linagora.tmail.api;
+
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Optional;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.linagora.tmail.HttpUtils;
+import com.linagora.tmail.OpenPaasConfiguration;
+
+import io.netty.handler.codec.http.HttpResponseStatus;
+import reactor.core.publisher.Mono;
+import reactor.netty.ByteBufMono;
+import reactor.netty.http.client.HttpClient;
+import reactor.netty.http.client.HttpClientResponse;
+
+public class OpenPaasRestClient {
+    private static final Duration RESPONSE_TIMEOUT = Duration.ofSeconds(10);
+    private static final String AUTHORIZATION_HEADER = "Authorization";
+    private final HttpClient client;
+    private final ObjectMapper deserializer = new ObjectMapper();
+
+    public OpenPaasRestClient(OpenPaasConfiguration openPaasConfiguration) {
+        URL apiUrl = openPaasConfiguration.restClientUrl();
+        String user = openPaasConfiguration.restClientUser();
+        String password = openPaasConfiguration.restClientPassword();
+        this.client = HttpClient.create()
+            .baseUrl(apiUrl.toString())
+            .headers(headers -> headers.add(AUTHORIZATION_HEADER, HttpUtils.createBasicAuthenticationToken(user, password)))
+            .responseTimeout(RESPONSE_TIMEOUT);
+    }
+
+    public Mono<Optional<OpenPaasUserResponse>> getUserById(String openPaasUserId) {
```

Comment:
```suggestion
    public Mono<OpenPaasUserResponse> getUserById(String openPaasUserId) {
```

Mono itself can represent emptyness we can strip that Optional.

Reply:
Tried it, and things got complicated. The second `.switchIfEmpty` is called even when status code is 404. I'm keeping the optional unless there is a better way.

```java
Mono.just(httpClientResponse.status())
            .filter(httpStatus -> httpStatus.equals(HttpResponseStatus.OK))
            .flatMap(httpStatus -> dataBuf.asByteArray())
            .map(this::parseUserResponse)
            .onErrorResume(e -> Mono.error(new OpenPaasRestClientException("Bad user response body format", e)))
            .switchIfEmpty(Mono.just(httpClientResponse.status())
                .filter(HttpResponseStatus.NOT_FOUND::equals)
                .flatMap(httpNotFound -> Mono.empty()))
            .switchIfEmpty(dataBuf.asString(StandardCharsets.UTF_8)
                .switchIfEmpty(Mono.just(""))
                .flatMap(errorResponse -> Mono.error(new OpenPaasRestClientException(
                    String.format("""
                            Error when getting OpenPaas user response.\s
                            Response Status = %s,
                            Response Body = %s""", statusCode, errorResponse)))));
```
for context: https://github.com/reactor/reactor-core/issues/1186#issuecomment-505575896

Reply:
So pro tips: do not go to early into a reactive pipeline and handle error handling prior to it

I'm pretty sure:

```
switch (response) {
   case HttpResponseStatus.OK -> return dataBuf.asByteArray()
            .map(this::parseUserResponse)
            .onErrorResume(e -> Mono.error(new OpenPaasRestClientException("Bad user response body format", e)))
   case HttpResponseStatus.NOT_FOUND ->
       LOGGER.info(...)
       return Mono.empty();
    default -> Mono.error(new OpenPaasRestClientException(
                    String.format("""
                            Error when getting OpenPaas user response.\s
                            Response Status = %s,
                            Response Body = %s""", statusCode, errorResponse))))
}
```

Or something like this would go a long long way.